system/jlib/jptree.cpp

```diff
@@ -531,6 +541,377 @@ const char *queryHead(const char *xpath, StringBuffer &head)
+{
+    /*
+    * NB: IPT's are not thread safe. It is up to the caller to ensure multiple writers do not contend.
+    * ( Dali for example ensures writer threasds are exclusive )
+    * 
+    * That means multiple reader threads could be here concurrently.
+    * >1 could be constructing the qualifier map for the 1st time.
+    * For new attr updates (where map already exists), it will block on map::crit,
+    * so that there is at most 1 thread updating the map. The underlying unordered_multiset
+    * is thread safe if 1 writer, and multiple readers.
+    * 
+    * On initial map creation, allow concurrency, but only 1 will succeed to swap in the new active map.
+    * That could mean a new attr/prop. mapping is lost, until next used.
+    * NB: once the map is live, updates are write ops. and so, just as with the IPT 
+    * itself, it is expected that something will keep it thread safe (as Dali does)
+    * 
+    */
+
+    // NB: only support simple @<attrname>=<value> qualifiers
+
+    if (((unsigned)-1) == siblingMapThreshold) // disabled
+        return nullptr;
+
+    PTree &_child = (PTree &)child;
+    if (child.isCaseInsensitive()) // NB: could support but not worth it.
+        return nullptr;
+
+    IPTArrayValue *value = _child.queryValue();
+    if (!value)
+        return nullptr;
+    CQualifierMap *map = value->queryMap();
+    if (!map)
+    {
+        if (!value->isArray() || (value->elements() < siblingMapThreshold))
+            return nullptr;
+    }
+
+    unsigned lhsLen, rhsLen;
+    const char *rhsStart;
+    const char *xpath = xxpath;
+    if (!parseEqualityQualifier(xpath, lhsLen, rhsStart, rhsLen))
+        return nullptr;
+    MAKE_LSTRING(lhs, xxpath, lhsLen);
+    MAKE_LSTRING(rhs, rhsStart, rhsLen);
+
+    // NB: there can be a race here where >1 reader is constructing new map
+    CValueMap *valueMap = nullptr;
+    if (map)
+        valueMap = map->addMappingIfNew(lhs, *value);
+    else
+    {
+        OwnedPtr<CQualifierMap> newMap = new CQualifierMap();
+        valueMap = newMap->addMapping(lhs, *value);
+
+        /*
+        * NB: it's possible another read thread got here 1st, and swapped in a map.
+        * setMap returns the existing map, and the code below checks to see if it already
+        * handles the 'lhs' we're adding, if it doesn't it re-adds the qualifier mappings.
+        */
+        map = value->setMap(newMap);
+        if (!map) // successfully swapped newMap in.
+            map = newMap.getClear(); // NB: setMap owns
+        else // another thread has swapped in a map whilst I was creating new one
+            valueMap = map->addMappingIfNew(lhs, *value);
+    }
+
+    xxpath = xpath; // update parsed position
+
+    auto range = valueMap->find(rhs);
+    if (range.first != range.second)
+        return new CMapQualifierIterator(*map, range.first, range.second);
+    else
+        return LINK(nullPTreeIterator);
```

Comment:
unrelated, but nullPTreeIterator could implement Link and release as no-ops.  It would avoid some atomic operations.